### PR TITLE
Require PHP 5.5+

### PR DESCRIPTION
**Pull request by Alkarex**
https://github.com/FreshRSS/FreshRSS/issues/2469#issuecomment-522255093
I think it would be reasonable to require PHP 5.5+ for the core of FreshRSS after all.

As Frenzie said, WordPress currently requires PHP 5.6.20+, and it is the most popular PHP application.

We would loose about 20% of the PHP servers according to https://w3techs.com/technologies/details/pl-php/5/all but I expect this number to drop fast after the release of CentOS 8 (CentOS accounts for 17% of Linux servers https://w3techs.com/technologies/details/os-linux/all/all ).

Distributions:
* no impact on Ubuntu, Fedora, Alpine, OpenWRT, FreeBSD, OpenSuze, Mageia, as all active versions have PHP ≥ 7
* no impact on Synology, as all active versions have PHP ≥ 5.6
* we drop Debian 8 Jessie (-2020) - we keep supporting Debian 9 Stretch (2017-06) - current is Debian 10 Buster
* we drop Red Hat 7 (-2024) - we keep supporting RHEL 8 (2019-05)
* we drop CentOS 7 (-2024) - we will support CentOS 8 (to be released soonish)

When dropping older versions, I can better like when it is for a good reason, and there is actually one with PHP 5.5, namely generators (yield) https://php.net/language.generators.overview which I consider
using.

### CODE REVIEW

**Comment by Alkarex** [comment]
Previous discussions:
* https://github.com/FreshRSS/FreshRSS/pull/1618
* https://github.com/FreshRSS/FreshRSS/issues/789#issuecomment-73878076

File: lib/lib_rss.php [hunk 1]
```diff
@@ -1,9 +1,9 @@
 <?php
-if (version_compare(PHP_VERSION, '5.3.8', '<')) {
-	die('FreshRSS error: FreshRSS requires PHP 5.3.8+!');
+if (version_compare(PHP_VERSION, '5.5.0', '<')) {
+	die('FreshRSS error: FreshRSS requires PHP 5.5.0+!');
 }
 
-if (!function_exists('json_decode')) {
+if (!function_exists('json_decode')) {	//PHP bug #63520 < PHP 7
```
Comment (Alkarex) [inline lib/lib_rss.php:6]: https://github.com/FreshRSS/FreshRSS/issues/306

File: .travis.yml [hunk 2]
```diff
@@ -30,12 +30,7 @@ env:
 matrix:
   fast_finish: true
   include:
-    # PHP 5.3 only runs on Ubuntu 12.04 (precise), not 14.04 (trusty)
-    - php: "5.3"
-      dist: precise
-    # PHP 5.4 & 5.5 only run on Travis in 14.04 (trusty), not 16.04 (xenial)
-    - php: "5.4"
-      dist: trusty
+    # PHP 5.5 only run on Travis in 14.04 (trusty), not 16.04 (xenial)
```
Comment (Frenzie) [inline .travis.yml]:
```suggestion
    # PHP 5.5 only runs on Travis in 14.04 (trusty), not 16.04 (xenial)
```
Stupid nit but the old sentence referred to two versions. ;-)

File: README.fr.md [hunk 3]
```diff
@@ -219,7 +219,6 @@ Tout client supportant une API de type Fever ; Sélection :
 * [bcrypt.js](https://github.com/dcodeIO/bcrypt.js)
 * [phpQuery](https://github.com/phpquery/phpquery)
 * [Services_JSON](https://pear.php.net/pepr/pepr-proposal-show.php?id=198)
-* [password_compat](https://github.com/ircmaxell/password_compat)
```
Comment (Frenzie) [inline README.fr.md:222]: I don't think `git rm lib/password_compat.php` is part of this PR yet?

Comment (Frenzie) [reply README.fr.md:222]: My bad, GH hides big changes so at a glance it looks like there are none.

Comment (Alkarex) [reply README.fr.md:222]: Yes, `password_compat` is removed because it is not needed from PHP 5.5+

**Comment by Alkarex** [comment]
Actually, moving to PHP 5.6+ would not be much harder than 5.5+ based on the statistics and versions available in the distributions (which have either a version <= 5.4 or >= 5.6).
I could not find any big distribution on DistroWatch that is maintaining a PHP 5.5, e.g. https://distrowatch.com/table.php?distribution=Debian

**Comment by Frenzie** [comment]
No real objections here. Anything specific though?

https://www.php.net/manual/en/migration56.new-features.php

A few of those sound somewhat interesting (e.g., variadic functions) but at the same time a minor syntactic convenience isn't much of a reason.

**Comment by Alkarex** [comment]
Ok. So let's stick to 5.5 now with a note that if we need 5.6 for anything, we can update without delay.

**Comment by marienfressinaud** [comment]
I'm ok with this decision too :+1:

I don't know how relevant the stats from w3techs.com are for our users, it may be interesting to do a survey to ask users where and how they host their instance of FreshRSS. Or even to add an  option in FRSS (disabled by default) to send anonymous information automatically (but it's more work and I don't have time to work on this for the moment)

File: lib/Minz/Configuration.php [hunk 4]
```diff
@@ -198,7 +198,7 @@ public function save() {
 			return false;
 		}
 
-		// Clear PHP 5.5+ cache for include
+		// Clear PHP cache for include
 		if (function_exists('opcache_invalidate')) {
```
Comment (marienfressinaud) [inline lib/Minz/Configuration.php:202]: Do we still need to check if the function exists?

**Review by marienfressinaud — APPROVED** [review]

File: lib/Minz/Configuration.php (hunk 4 above)
Comment (Alkarex) [reply lib/Minz/Configuration.php:202]: Some distributions do not have `opcache` by default, e.g. https://pkgs.alpinelinux.org/packages?name=php*opcache*&branch=edge&arch=x86_64 , which makes me realise that:
* We should add that do the documentation
* We should add it in the Dockerfile for Alpine
* Maybe that might explain a part of why Alpine was so much slower than Ubuntu in my last test https://github.com/FreshRSS/FreshRSS/pull/2205

Comment (Alkarex) [reply lib/Minz/Configuration.php:202]: https://github.com/FreshRSS/FreshRSS/pull/2498

**Comment by Alkarex** [comment]
Actually, I am refactoring a bit our PDO code (https://github.com/FreshRSS/FreshRSS/pull/2522 ) and I do have a wish to move to PHP 5.6+ to be able to use constant arrays
https://www.php.net/manual/en/migration56.new-features.php

Instead of the current `global` approach, which is not so nice, e.g.:
https://github.com/FreshRSS/FreshRSS/blob/8dcdde6251ae4dfc690b1a014488df125c5e5cdc/app/Models/UserDAO.php#L21-L22

https://github.com/FreshRSS/FreshRSS/blob/c76a318193cda63064625b2d92c719b7150d7d64/app/SQL/install.sql.pgsql.php#L4-L9

What do you think? Is it worth it?

**Comment by Frenzie** [comment]
It's a real feature, so I don't see why not. :+1:

**Comment by marienfressinaud** [comment]
It would be a nice simplification :) and https://github.com/FreshRSS/FreshRSS/pull/2495#issuecomment-522538468 was already a sufficient reason to move on

**Comment by Alkarex** [comment]
CentOS 8 will be released next week with PHP 7.2 https://wiki.centos.org/About/Building_8

**Pull request merged**